src/m365/spo/commands/folder/folder-retentionlabel-ensure.ts

```diff
@@ -0,0 +1,174 @@
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    if (this.verbose) {
+      logger.logToStderr(`Applying retention label to folder ${args.options.folderId || args.options.folderUrl} in site at ${args.options.webUrl}...`);
+    }
+    try {
+      const folderProperties = await this.getFolderProperties(args);
+      if (typeof folderProperties !== 'string') {
+        const options: SpoListItemRetentionLabelEnsureCommandOptions = {
+          webUrl: args.options.webUrl,
+          listId: folderProperties.listId,
+          listItemId: folderProperties.id,
+          name: args.options.name,
+          output: 'json',
+          debug: this.debug,
+          verbose: this.verbose
+        };
+
+        const spoListItemRetentionLabelEnsureCommandOutput = await Cli.executeCommandWithOutput(SpoListItemRetentionLabelEnsureCommand as Command, { options: { ...options, _: [] } });
+        if (this.verbose) {
+          logger.logToStderr(spoListItemRetentionLabelEnsureCommandOutput.stderr);
+        }
+      }
+      else {
+        const options: SpoListRetentionLabelEnsureCommandOptions = {
+          webUrl: args.options.webUrl,
+          listUrl: folderProperties as string,
+          name: args.options.name,
+          output: 'json',
+          debug: this.debug,
+          verbose: this.verbose
+        };
+
+        const spoListRetentionLabelEnsureCommandOutput = await Cli.executeCommandWithOutput(SpoListRetentionLabelEnsureCommand as Command, { options: { ...options, _: [] } });
+        if (this.verbose) {
+          logger.logToStderr(spoListRetentionLabelEnsureCommandOutput.stderr);
+        }
+      }
+    }
+    catch (err: any) {
+      this.handleRejectedODataJsonPromise(err);
+    }
+  }
+
+  private async getFolderProperties(args: CommandArgs): Promise<folderOptions | string> {
+    const requestOptions: AxiosRequestConfig = {
+      headers: {
+        'accept': 'application/json;odata=nometadata'
+      },
+      responseType: 'json'
+    };
+
+    let requestUrl = `${args.options.webUrl}/_api/web/`;
+
+    if (args.options.folderId) {
+      requestUrl += `GetFolderById('${args.options.folderId}')`;
+    }
+    else {
+      const serverRelativeUrl = urlUtil.getServerRelativePath(args.options.webUrl, args.options.folderUrl!);
+      requestUrl += `GetFolderByServerRelativeUrl('${formatting.encodeQueryParameter(serverRelativeUrl)}')`;
+    }
+
+    requestOptions.url = `${requestUrl}?$expand=ListItemAllFields,ListItemAllFields/ParentList&$select=ServerRelativeUrl,ListItemAllFields/ParentList/Id,ListItemAllFields/Id`;
+
+    const response = await request.get<FolderProperties>(requestOptions);
+
+    // if List, return the server relative list url because Unique ID does not work for api /list(guid 'Unique ID') in command list retentionlabel ensure
+    if (!response.ListItemAllFields) {
+      return response.ServerRelativeUrl;
```

Comment:
Hi @nicodecleyre, I don't think its a good choice here to return a string or an object. 
Let's just the FolderProperties response in both situations, just like `FileProperties.ts` is used.

```typescript
return response; //as this is a FolderProperties 
```

and use the `if(folderProperties.ListItemAllFields != undefined)` to check what command to call.

Reply:
That's a better way to handle the 2 different scenario's! Updated the pr!